### ISSUE TITLE
Can't get needed information from com.oracle.svm.util.ModuleSupport for observed exception

### ISSUE THREAD

**Issue by andrejusc** [oracle/graal#4804]
**Describe the issue**
Trying to invoke `native-image` generation with (besides other) options `-p target/service.jar:target/libs -m xyz/MyClass` leads to such exception:
```
Fatal error: com.oracle.svm.util.ModuleSupport$ModuleSupportError: open of packages from module jersey.client to class com.oracle.svm.hosted.ClassLoaderSupportImpl failed. No module named jersey.client in boot layer.
	at org.graalvm.nativeimage.base/com.oracle.svm.util.ModuleSupport.accessPackagesToClass(ModuleSupport.java:116)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ClassLoaderSupportImpl.getResourceBundle(ClassLoaderSupportImpl.java:241)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.jdk.localization.LocalizationFeature.prepareBundle(LocalizationFeature.java:538)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ResourcesFeature$ResourcesRegistryImpl.lambda$addResourceBundles$4(ResourcesFeature.java:167)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ConditionalConfigurationRegistry.registerConditionalConfiguration(ConditionalConfigurationRegistry.java:43)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ResourcesFeature$ResourcesRegistryImpl.addResourceBundles(ResourcesFeature.java:167)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ResourceConfigurationParser.parseBundle(ResourceConfigurationParser.java:110)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ResourceConfigurationParser.parseTopLevelObject(ResourceConfigurationParser.java:93)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ResourceConfigurationParser.parseAndRegister(ResourceConfigurationParser.java:51)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ConfigurationParser.parseAndRegister(ConfigurationParser.java:71)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.config.ConfigurationParserUtils.doParseAndRegister(ConfigurationParserUtils.java:125)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.config.ConfigurationParserUtils.lambda$parseAndRegisterConfigurations$3(ConfigurationParserUtils.java:111)
	at java.base/java.util.stream.ReferencePipeline$4$1.accept(ReferencePipeline.java:214)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.config.ConfigurationParserUtils$1.tryAdvance(ConfigurationParserUtils.java:104)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:310)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.IntPipeline.reduce(IntPipeline.java:515)
	at java.base/java.util.stream.IntPipeline.sum(IntPipeline.java:473)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.config.ConfigurationParserUtils.parseAndRegisterConfigurations(ConfigurationParserUtils.java:113)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ResourcesFeature.beforeAnalysis(ResourcesFeature.java:186)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$9(NativeImageGenerator.java:722)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:78)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:722)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:564)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:521)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:407)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:585)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
```
Separately checking presence of `jersey.client` module within `target/libs` folder returns such:
```
$ java -p target/libs --list-modules | grep "jersey.client"
jersey.client@2.35 file:///dev/myproject/target/libs/jersey-client-2.35.jar automatic
```

And then checking source of `com.oracle.svm.util.ModuleSupport.accessPackagesToClass` method - there is no Debug output inside it to be printed if verbose mode is enabled. It would be handy though to print what modules are resolved for this `ModuleLayer.boot()` layer/etc. to understand issue root cause.

**Steps to reproduce the issue**
This happens at such stage of native-image generation:
```
[1/7] Initializing...                                                                                   (19.6s @ 1.11GB)
 Version info: 'GraalVM 22.2.0 Java 17 EE'
 Java version info: '17.0.4+11-LTS-jvmci-22.2-b05'
 C compiler: x86_64-linux-musl-gcc (redhat, x86_64, 8.5.0)
 Garbage collector: Serial GC
 1 user-specific feature(s)
 - com.oracle.svm.thirdparty.gson.GsonFeature

Fatal error: com.oracle.svm.util.ModuleSupport$ModuleSupportError: open of packages from module jersey.client to class com.oracle.svm.hosted.ClassLoaderSupportImpl failed. No module named jersey.client in boot layer.
...
```

**Describe GraalVM and your environment:**
```
$ native-image --version
GraalVM 22.2.0 Java 17 EE (Java Version 17.0.4+11-LTS-jvmci-22.2-b05)
```

**Comment by oubidar-Abderrahim** [comment]
Hi, Thank you for reporting this, could you please share a reproducer for this issue? Thank you

**Comment by andrejusc** [comment]
I'm trying to prepare isolated use case but now facing this strange case:
Invoking this way - getting as expected reference to use help
```
$ native-image --features
Error: Please specify class containing the main entry point method. (see --help)
```

but then invoking this way:
```
$ native-image --features org.example.Abc
Error: Unrecognized option: --features
```
Any clue on such behavior? Or should I raise another issue on this as well?

**Comment by andrejusc** [comment]
@oubidar-Abderrahim - I've got a bit closer to my use case, but still not there and more questions coming. Attached is `hellonative20220815.tgz` archive file. Untar it and initially run inside `hellonative` folder:
```
mvn clean package
```
That will produce inside `nativemain` folder such file `target/service.jar` and will populate `target/libs` folder with needed dependencies. Next:
```
cd nativemain
native-image --features=org.example.feature.graal.HelloNativeFeature --add-opens jersey.client/org.glassfish.jersey.client=org.graalvm.nativeimage.builder --static -H:Name=service -H:IncludeResources="META-INF/native-image/native-image.properties" -p target/service.jar:target/libs -m nativemain/org.example.main.HelloNative
```
Note that in above command I have explicit `--add-opens jersey.client/org.glassfish.jersey.client=org.graalvm.nativeimage.builder` part

In output of native-image generation there will be such (I'm using GraalVM JDK17 22.2) lines:
```
*** HelloNativeFeature ModuleLayer.boot().modules().size(): 63
*** HelloNativeFeature ModuleLayer.boot().modules(): [module jdk.security.auth...
```
but no presence of `jersey.client` as a module within those 63 modules output, while I'd expect it to show up somewhere.

Please let me know if instead of `ModuleLayer.boot()` I should explore some other object/etc., which may present other modules at runtime.

[hellonative20220815.tgz](https://github.com/oracle/graal/files/9337678/hellonative20220815.tgz)
.

**Comment by andrejusc** [comment]
@oubidar-Abderrahim - I was able now to prepare isolated use case with exact same error.

Initial steps to repro are same as in above comment. Please let me know if you able to reproduce using my new archive.

And then during native-image generation and this above comment's command:
```
native-image --features=org.example.feature.graal.HelloNativeFeature --verbose --add-opens jersey.client/org.glassfish.jersey.client=org.graalvm.nativeimage.builder --static -H:Name=service -H:IncludeResources="META-INF/native-image/native-image.properties" -p target/service.jar:target/libs -m nativemain/org.example.main.HelloNative
```
you'll see:
```
[1/7] Initializing...                                                                    (7.4s @ 0.23GB)
 Version info: 'GraalVM 22.2.0 Java 17 EE'
 Java version info: '17.0.4+11-LTS-jvmci-22.2-b05'
 C compiler: gcc (redhat, x86_64, 8.5.0)
 Garbage collector: Serial GC
[hellonative20220816.tgz](https://github.com/oracle/graal/files/9351334/hellonative20220816.tgz)

 1 user-specific feature(s)
 - org.example.feature.graal.HelloNativeFeature

Fatal error: com.oracle.svm.util.ModuleSupport$ModuleSupportError: open of packages from module jersey.client to class com.oracle.svm.hosted.ClassLoaderSupportImpl failed. No module named jersey.client in boot layer.
	at org.graalvm.nativeimage.base/com.oracle.svm.util.ModuleSupport.accessPackagesToClass(ModuleSupport.java:116)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ClassLoaderSupportImpl.getResourceBundle(ClassLoaderSupportImpl.java:241)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.jdk.localization.LocalizationFeature.prepareBundle(LocalizationFeature.java:538)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ResourcesFeature$ResourcesRegistryImpl.lambda$addResourceBundles$4(ResourcesFeature.java:167)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ConditionalConfigurationRegistry.registerConditionalConfiguration(ConditionalConfigurationRegistry.java:43)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ResourcesFeature$ResourcesRegistryImpl.addResourceBundles(ResourcesFeature.java:167)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ResourceConfigurationParser.parseBundle(ResourceConfigurationParser.java:110)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ResourceConfigurationParser.parseTopLevelObject(ResourceConfigurationParser.java:93)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ResourceConfigurationParser.parseAndRegister(ResourceConfigurationParser.java:51)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ConfigurationParser.parseAndRegister(ConfigurationParser.java:71)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.config.ConfigurationParserUtils.doParseAndRegister(ConfigurationParserUtils.java:125)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.config.ConfigurationParserUtils.lambda$parseAndRegisterConfigurations$3(ConfigurationParserUtils.java:111)
	at java.base/java.util.stream.ReferencePipeline$4$1.accept(ReferencePipeline.java:214)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.config.ConfigurationParserUtils$1.tryAdvance(ConfigurationParserUtils.java:104)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:310)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.IntPipeline.reduce(IntPipeline.java:515)
	at java.base/java.util.stream.IntPipeline.sum(IntPipeline.java:473)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.config.ConfigurationParserUtils.parseAndRegisterConfigurations(ConfigurationParserUtils.java:113)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ResourcesFeature.beforeAnalysis(ResourcesFeature.java:186)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$9(NativeImageGenerator.java:722)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:78)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:722)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:564)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:521)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:407)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:585)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)

```
I've added also extra output for 3 Modules layers, i.e. `boot` layer, one layer that belongs to `org.graalvm.nativeimage.builder` module and one that belongs to my `nativefeature` module.

Partial output is like such for those different layers:
```
*** HelloNativeFeature ModuleLayer.boot().modules().size(): 63
*** HelloNativeFeature ModuleLayer.boot().modules(): [module jdk.internal.opt, module java.rmi, module jdk.compiler, ...
```
```
*** niBuilder module: module org.graalvm.nativeimage.builder
*** niBuilder.getLayer().modules().size(): 63
*** niBuilder.getLayer().modules(): [module jdk.internal.opt, module java.rmi, module jdk.compiler, ...
```
```
*** nativefeature.getLayer().modules().size(): 12
*** nativefeature.getLayer().modules(): [module nativemain, module java.annotation, module org.graalvm.nativeimage.librarysupport, module jakarta.activation, module nativefeature, module java.ws.rs, module jakarta.inject, module osgi.resource.locator, module jersey.common, module graal.microservices, module jersey.client, module java.xml.bind]
```

**Comment by olpaw** [comment]
We currently do not support the `target/libs`-part of `-p target/service.jar:target/libs`. If you replace `target/libs` with `target/libs/*.jar` `--add-opens=...` will get processed as expected, but that's still not enough as it is missing module `nativefeature` now. For some reason `target/libs/*.jar` is not processed the same as `target/libs`.

**Comment by andrejusc** [comment]
@olpaw - appreciated your update. May you have any optimistic ETA on such support addition as such is blocking one of our projects? 

Maybe you could also improve that error message:
```
Fatal error: com.oracle.svm.util.ModuleSupport$ModuleSupportError: open of packages from module jersey.client to class com.oracle.svm.hosted.ClassLoaderSupportImpl failed. No module named jersey.client in boot layer.
```

to something like this to indicate which package (see end of long line below) was passed (I could see that in related source code):
```
Fatal error: com.oracle.svm.util.ModuleSupport$ModuleSupportError: open of packages from module jersey.client to class com.oracle.svm.hosted.ClassLoaderSupportImpl failed. No module named jersey.client in boot layer. **Package requested was: abc.xyz**
```

Please also note that my usage of `--add-opens` here in generation command line was an attempt to see if I could make it work without that Fatal error and while analyzing Module layers via my introduced output I got to know that my `nativefeature` module located within layer, which has parent layer containing `org.graalvm.nativeimage.builder`

There is such output produced by my demo code:
```
*** nativefeature.getLayer().parents(): [java.management.rmi, org.graalvm.nativeimage.pointsto, jdk.unsupported, jdk.jpackage, java.sql, jdk.security.jgss, java.transaction.xa, java.prefs, jdk.internal.opt, org.graalvm.truffle, jdk.management, jdk.internal.vm.compiler.management, org.graalvm.sdk, java.security.sasl, jdk.naming.rmi, java.net.http, org.graalvm.locator, jdk.jdeps, com.oracle.svm.shadowed.org.bytedeco.javacpp, java.instrument, com.oracle.svm.svm_enterprise.llvm, java.base, jdk.management.jfr, jdk.sctp, jdk.jfr, jdk.security.auth, com.oracle.graal.graal_enterprise, java.rmi, jdk.internal.vm.ci, jdk.javadoc, java.xml, java.scripting, jdk.unsupported.desktop, jdk.charsets, java.compiler, java.naming, jdk.crypto.cryptoki, com.oracle.truffle.truffle_nfi, org.graalvm.nativeimage.llvm, jdk.naming.dns, jdk.crypto.ec, java.smartcardio, java.security.jgss, java.logging, com.oracle.svm.shadowed.org.bytedeco.javacpp.linux.x86_64, java.datatransfer, java.xml.crypto, org.graalvm.nativeimage.base, jdk.jartool, com.oracle.svm.shadowed.org.bytedeco.llvm, jdk.internal.vm.compiler.truffle.jfr, jdk.internal.vm.compiler, jdk.compiler, java.management, com.oracle.svm.shadowed.org.bytedeco.llvm.linux.x86_64, jdk.zipfs, com.oracle.svm.svm_enterprise, org.graalvm.nativeimage.objectfile, java.desktop, org.graalvm.nativeimage.builder, jdk.localedata, jdk.random, jdk.jlink]
```

and you could see in it presence of `org.graalvm.nativeimage.builder`

So additional question here is if usage of above `--add-opens` is needed in general as to my understanding via parent-child Module layers relation visibility should be potentially in place. But maybe I'm wrong in that.

**Comment by olpaw** [comment]
```
Fatal error: com.oracle.svm.util.ModuleSupport$ModuleSupportError: open of packages from module jersey.client to class com.oracle.svm.hosted.ClassLoaderSupportImpl failed. No module named jersey.client in boot layer.
```
@andrejusc that is to be expected. Why would `jersey.client` be in the boot module-layer of the image builder? It can only be in the module-layer that the image builder creates for the `-p target/service.jar:target/libs` argument that gets passed to it. See `com.oracle.svm.hosted.NativeImageClassLoaderSupport#moduleLayerForImageBuild`.

From your feature `org.example.feature.graal.HelloNativeFeature` you can get to that module-layer with
```java
HelloNativeFeature.class.getModule().getLayer().modules()
```
Also I see that in
```
hellonative20220816/nativefeature/src/main/resources/META-INF/native-image/native-image.properties
```
you have
```
Args=\
  --features=org.example.HelloNativeFeature
```
This makes no sense. It should be:
```
Args=\
  --features=org.example.feature.graal.HelloNativeFeature
```
Then you can remove `--features=org.example.feature.graal.HelloNativeFeature` because the `native-image.properties` file will take care of that.

**Comment by olpaw** [comment]
With the changes above I get to the following:
```
$ native-image --add-opens jersey.client/org.glassfish.jersey.client=org.graalvm.nativeimage.builder --static -H:Name=service -H:IncludeResources="META-INF/native-image/native-image.properties" -p target/service.jar$(for jar in target/libs/*.jar; do echo -n :$jar; done) -m nativemain/org.example.main.HelloNative 
========================================================================================================================
GraalVM Native Image: Generating 'service' (static executable)...
========================================================================================================================
*** HelloNativeFeature.class.getModule().getLayer().modules(): [module jersey.client, module jersey.common, module jakarta.activation, module java.xml.bind, module org.graalvm.nativeimage.librarysupport, module osgi.resource.locator, module java.annotation, module java.ws.rs, module graal.microservices, module nativefeature, module nativemain, module jakarta.inject]
[1/7] Initializing...                                                                                    (2.7s @ 0.24GB)
 Version info: 'GraalVM 22.2.0 Java 17 EE'
 Java version info: '17.0.4+11-LTS-jvmci-22.2-b05'
 C compiler: gcc (redhat, x86_64, 12.1.1)
 Garbage collector: Serial GC
 1 user-specific feature(s)
 - org.example.feature.graal.HelloNativeFeature

Fatal error: com.oracle.svm.util.ModuleSupport$ModuleSupportError: open of packages from module jersey.client to class com.oracle.svm.hosted.ClassLoaderSupportImpl failed. No module named jersey.client in boot layer.
	at org.graalvm.nativeimage.base/com.oracle.svm.util.ModuleSupport.accessPackagesToClass(ModuleSupport.java:116)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ClassLoaderSupportImpl.getResourceBundle(ClassLoaderSupportImpl.java:241)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.jdk.localization.LocalizationFeature.prepareBundle(LocalizationFeature.java:538)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ResourcesFeature$ResourcesRegistryImpl.lambda$addResourceBundles$4(ResourcesFeature.java:167)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ConditionalConfigurationRegistry.registerConditionalConfiguration(ConditionalConfigurationRegistry.java:43)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ResourcesFeature$ResourcesRegistryImpl.addResourceBundles(ResourcesFeature.java:167)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ResourceConfigurationParser.parseBundle(ResourceConfigurationParser.java:110)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ResourceConfigurationParser.parseTopLevelObject(ResourceConfigurationParser.java:93)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ResourceConfigurationParser.parseAndRegister(ResourceConfigurationParser.java:51)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.configure.ConfigurationParser.parseAndRegister(ConfigurationParser.java:71)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.config.ConfigurationParserUtils.doParseAndRegister(ConfigurationParserUtils.java:125)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.config.ConfigurationParserUtils.lambda$parseAndRegisterConfigurations$3(ConfigurationParserUtils.java:111)
	at java.base/java.util.stream.ReferencePipeline$4$1.accept(ReferencePipeline.java:214)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.config.ConfigurationParserUtils$1.tryAdvance(ConfigurationParserUtils.java:104)
...
```
As you can see we are now running into another issue. This is module-related bug in resource bundle handling that will also need to be fixed.

**Comment by olpaw** [comment]
> @olpaw - appreciated your update. May you have any optimistic ETA on such support addition as such is blocking one of our projects?

For now you can use the workaround I have shown above. Instead of
```
-p target/service.jar:target/libs
```
use
```
-p target/service.jar$(for jar in target/libs/*.jar; do echo -n :$jar; done)
```

I should have that fixed on master by the end of next week.

> So additional question here is if usage of above --add-opens is needed in general as to my understanding via parent-child Module layers relation visibility should be potentially in place. But maybe I'm wrong in that.

The explicit `--add-opens` should not be needed at all. Your use if `RuntimeReflection.register` in `org.example.feature.graal.HelloNativeFeature` should take care of opening up stuff as needed.

The other issue I have shown in https://github.com/oracle/graal/issues/4804#issuecomment-1219238428 will likely also be fixed in the course of next week. I suspect it comes from
```
hellonative20220816/nativemain/src/main/resources/META-INF/native-image/resource-config.json
```
there we have
```
{
  "resources":{
  "includes":[]},
  "bundles":[
    {
      "name":"org.glassfish.jersey.client.internal.localization",
      "locales":[""]
    }
  ]
}
```
This bundle is not located in a module that is part of the boot module-layer and the LocalizationFeature seems to assume it is. This needs to be fixed.

**Comment by andrejusc** [comment]
Thank you for this advice:
```
This makes no sense. It should be:

Args=\
  --features=org.example.feature.graal.HelloNativeFeature
```
this was on my side after rushing to prepare isolated use case and renaming packages. Corrected now.

So I changed that inside my native-image.properties file and trying to invoke now with your workaround suggestion:
```
$ native-image --verbose --static -H:Name=service --add-opens jersey.client/org.glassfish.jersey.client=org.graalvm.nativeimage.builder -p target/service.jar$(for jar in target/libs/*.jar; do echo -n :$jar; done) -m nativemain/org.example.main.HelloNative
```
but such leads so far to same original error:
```
[1/7] Initializing...                                                                            (6.6s @ 0.24GB)
 Version info: 'GraalVM 22.2.0 Java 17 EE'
 Java version info: '17.0.4+11-LTS-jvmci-22.2-b05'
 C compiler: gcc (redhat, x86_64, 8.5.0)
 Garbage collector: Serial GC
 1 user-specific feature(s)
 - org.example.feature.graal.HelloNativeFeature

Fatal error: com.oracle.svm.util.ModuleSupport$ModuleSupportError: open of packages from module jersey.client to class com.oracle.svm.hosted.ClassLoaderSupportImpl failed. No module named jersey.client in boot layer.
	at org.graalvm.nativeimage.base/com.oracle.svm.util.ModuleSupport.accessPackagesToClass(ModuleSupport.java:116)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ClassLoaderSupportImpl.getResourceBundle(ClassLoaderSupportImpl.java:241)
...
```
So potentially `LocalizationFeature` needs to be fixed first here to unblock things in general and then next one `target/libs` as folder support fix for `--module-path` reference.

**Comment by andrejusc** [comment]
Attaching here current latest code of isolated use case.
[hellonative20220818.tgz](https://github.com/oracle/graal/files/9372929/hellonative20220818.tgz)

Latest invocation attempted using this command:
```
native-image --verbose --static -H:Name=service --add-opens jersey.client/org.glassfish.jersey.client=org.graalvm.nativeimage.builder -p target/service.jar$(for jar in target/libs/*.jar; do echo -n :$jar; done) -m nativemain/org.example.main.HelloNative
```
and so far leading to same original issue.

**Comment by olpaw** [comment]
@andrejusc with the changes on PR https://github.com/oracle/graal/pull/4834 I'm able to build your example just fine:
```
hellonative20220816/nativemain$ native-image -o service -p target/service.jar:target/libs -m nativemain/org.example.main.HelloNative 
========================================================================================================================
GraalVM Native Image: Generating 'service' (executable)...
========================================================================================================================
[1/7] Initializing...                                                                                    (2.8s @ 0.21GB)
 Version info: 'GraalVM dev Java 17 CE'
 Java version info: '17.0.4+7-jvmci-22.3-b02'
 C compiler: gcc (redhat, x86_64, 12.1.1)
 Garbage collector: Serial GC
[2/7] Performing analysis...  [*****]                                                                    (5.5s @ 0.78GB)
   2,914 (73.03%) of  3,990 classes reachable
   3,518 (51.55%) of  6,825 fields reachable
  12,690 (42.34%) of 29,973 methods reachable
      28 classes,     0 fields, and   317 methods registered for reflection
      58 classes,    58 fields, and    52 methods registered for JNI access
       4 native libraries: dl, pthread, rt, z
[3/7] Building universe...                                                                               (1.1s @ 1.27GB)
[4/7] Parsing methods...      [*]                                                                        (0.7s @ 0.75GB)
[5/7] Inlining methods...     [***]                                                                      (0.3s @ 1.15GB)
[6/7] Compiling methods...    [**]                                                                       (3.8s @ 1.56GB)
[7/7] Creating image...                                                                                  (1.1s @ 2.00GB)
   4.11MB (36.35%) for code area:     7,283 compilation units
   6.94MB (61.43%) for image heap:  100,586 objects and 6 resources
 256.45KB ( 2.22%) for other data
  11.29MB in total
------------------------------------------------------------------------------------------------------------------------
Top 10 packages in code area:                               Top 10 object types in image heap:
 657.19KB java.util                                          917.72KB java.lang.String
 329.21KB java.lang                                          907.07KB byte[] for code metadata
 264.87KB java.text                                          849.22KB byte[] for general heap data
 216.35KB java.util.regex                                    638.84KB java.lang.Class
 192.83KB java.util.concurrent                               543.14KB java.util.HashMap$Node
 146.93KB java.math                                          536.64KB byte[] for java.lang.String
 116.63KB java.lang.invoke                                   233.59KB java.util.HashMap$Node[]
 105.05KB com.oracle.svm.core.genscavenge                    227.66KB com.oracle.svm.core.hub.DynamicHubCompanion
  95.93KB java.util.logging                                  163.29KB java.lang.String[]
  93.82KB java.util.stream                                   155.86KB java.util.concurrent.ConcurrentHashMap$Node
   1.89MB for 126 more packages                                1.57MB for 787 more object types
------------------------------------------------------------------------------------------------------------------------
                        0.4s (2.1% of total time) in 20 GCs | Peak RSS: 2.88GB | CPU load: 10.79
------------------------------------------------------------------------------------------------------------------------
Produced artifacts:
 /home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/service (executable)
 /home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/service.build_artifacts.txt (txt)
========================================================================================================================
Finished generating 'service' in 16.4s.
```
Running that image shows:
```
hellonative20220816/nativemain$ ./service 
=== Hello native!
=== ClientConfig: org.glassfish.jersey.client.ClientConfig@e817d48e
=== JerseyCaller: org.example.feature.JerseyCaller@1e5f20a6
```
I have also added an implementation of `--list-modules` to native-image. For your module-path argument we get:
```
hellonative20220816/nativemain$ native-image -p target/service.jar:target/libs --list-modules 
jakarta.activation file:///home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/target/libs/jakarta.activation-api-1.2.2.jar
jakarta.inject@2.6.1 file:///home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/target/libs/jakarta.inject-2.6.1.jar automatic
java.annotation@1.3.5 file:///home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/target/libs/jakarta.annotation-api-1.3.5.jar automatic
java.ws.rs file:///home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/target/libs/jakarta.ws.rs-api-2.1.6.jar
java.xml.bind@2.3.3 file:///home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/target/libs/jakarta.xml.bind-api-2.3.3.jar
jersey.client@2.35 file:///home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/target/libs/jersey-client-2.35.jar automatic
jersey.common@2.35 file:///home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/target/libs/jersey-common-2.35.jar automatic
nativefeature@1.0-SNAPSHOT file:///home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/target/libs/nativefeature-1.0-SNAPSHOT.jar
nativemain@1.0-SNAPSHOT file:///home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/target/service.jar
org.graalvm.nativeimage.librarysupport file:///home/pwoegere/OLabs/issues/GR-40456/graal/sdk/mxbuild/linux-amd64/GRAALVM_A5669547E3_JAVA17/graalvm-a5669547e3-java17-22.3.0-dev/lib/svm/library-support.jar
osgi.resource.locator@1.0.3 file:///home/pwoegere/OLabs/issues/GR-40456/hellonative20220816/nativemain/target/libs/osgi-resource-locator-1.0.3.jar automatic

com.oracle.svm.shadowed.org.bytedeco.javacpp file:///home/pwoegere/OLabs/issues/GR-40456/graal/sdk/mxbuild/linux-amd64/GRAALVM_A5669547E3_JAVA17/graalvm-a5669547e3-java17-22.3.0-dev/lib/svm/builder/javacpp-shadowed.jar
com.oracle.svm.shadowed.org.bytedeco.javacpp.linux.x86_64 file:///home/pwoegere/OLabs/issues/GR-40456/graal/sdk/mxbuild/linux-amd64/GRAALVM_A5669547E3_JAVA17/graalvm-a5669547e3-java17-22.3.0-dev/lib/svm/builder/javacpp-platform-specific-shadowed.jar open
com.oracle.svm.shadowed.org.bytedeco.llvm file:///home/pwoegere/OLabs/issues/GR-40456/graal/sdk/mxbuild/linux-amd64/GRAALVM_A5669547E3_JAVA17/graalvm-a5669547e3-java17-22.3.0-dev/lib/svm/builder/llvm-wrapper-shadowed.jar
com.oracle.svm.shadowed.org.bytedeco.llvm.linux.x86_64 file:///home/pwoegere/OLabs/issues/GR-40456/graal/sdk/mxbuild/linux-amd64/GRAALVM_A5669547E3_JAVA17/graalvm-a5669547e3-java17-22.3.0-dev/lib/svm/builder/llvm-platform-specific-shadowed.jar open
java.base@17.0.4
java.compiler@17.0.4
java.datatransfer@17.0.4
java.desktop@17.0.4
java.instrument@17.0.4
java.logging@17.0.4
java.management@17.0.4
java.management.rmi@17.0.4
java.naming@17.0.4
java.net.http@17.0.4
java.prefs@17.0.4
java.rmi@17.0.4
java.scripting@17.0.4
java.security.jgss@17.0.4
java.security.sasl@17.0.4
java.smartcardio@17.0.4
java.sql@17.0.4
java.transaction.xa@17.0.4
java.xml@17.0.4
java.xml.crypto@17.0.4
jdk.charsets@17.0.4
jdk.compiler@17.0.4
jdk.crypto.cryptoki@17.0.4
jdk.crypto.ec@17.0.4
jdk.httpserver@17.0.4
jdk.internal.opt@17.0.4
jdk.internal.vm.ci@17.0.4
jdk.internal.vm.compiler
jdk.internal.vm.compiler.management
jdk.internal.vm.compiler.truffle.jfr
jdk.jartool@17.0.4
jdk.javadoc@17.0.4
jdk.jdeps@17.0.4
jdk.jfr@17.0.4
jdk.jlink@17.0.4
jdk.jpackage@17.0.4
jdk.localedata@17.0.4
jdk.management@17.0.4
jdk.management.jfr@17.0.4
jdk.naming.dns@17.0.4
jdk.naming.rmi@17.0.4
jdk.random@17.0.4
jdk.sctp@17.0.4
jdk.security.auth@17.0.4
jdk.security.jgss@17.0.4
jdk.unsupported@17.0.4
jdk.unsupported.desktop@17.0.4
jdk.zipfs@17.0.4
org.graalvm.locator
org.graalvm.nativeimage.base file:///home/pwoegere/OLabs/issues/GR-40456/graal/sdk/mxbuild/linux-amd64/GRAALVM_A5669547E3_JAVA17/graalvm-a5669547e3-java17-22.3.0-dev/lib/svm/builder/native-image-base.jar
org.graalvm.nativeimage.builder file:///home/pwoegere/OLabs/issues/GR-40456/graal/sdk/mxbuild/linux-amd64/GRAALVM_A5669547E3_JAVA17/graalvm-a5669547e3-java17-22.3.0-dev/lib/svm/builder/svm.jar
org.graalvm.nativeimage.llvm file:///home/pwoegere/OLabs/issues/GR-40456/graal/sdk/mxbuild/linux-amd64/GRAALVM_A5669547E3_JAVA17/graalvm-a5669547e3-java17-22.3.0-dev/lib/svm/builder/svm-llvm.jar
org.graalvm.nativeimage.objectfile file:///home/pwoegere/OLabs/issues/GR-40456/graal/sdk/mxbuild/linux-amd64/GRAALVM_A5669547E3_JAVA17/graalvm-a5669547e3-java17-22.3.0-dev/lib/svm/builder/objectfile.jar
org.graalvm.nativeimage.pointsto file:///home/pwoegere/OLabs/issues/GR-40456/graal/sdk/mxbuild/linux-amd64/GRAALVM_A5669547E3_JAVA17/graalvm-a5669547e3-java17-22.3.0-dev/lib/svm/builder/pointsto.jar
org.graalvm.sdk
org.graalvm.truffle
```

As you can see, the module-layer that results from using `-p target/service.jar:target/libs` contains all the jars in `target/libs` and `target/service.jar` (in addition to the implicit `org.graalvm.nativeimage.librarysupport` module).
The parent module layer (below the first module layer) shows the image builder and the JDK modules.

**Comment by andrejusc** [comment]
@olpaw - is this now available inside some 22.2.x Dev build? We'll need to have this as some patch on top of 22.2.0.

**Comment by olpaw** [comment]
@andrejusc not yet. First the PR needs to get merged on master. Then the next https://github.com/graalvm/graalvm-ce-dev-builds will contain the changes.

**Comment by andrejusc** [comment]
@olpaw - in your above part of output:
```
[1/7] Initializing...                                                                                    (2.8s @ 0.21GB)
 Version info: 'GraalVM dev Java 17 CE'
 Java version info: '17.0.4+7-jvmci-22.3-b02'
 C compiler: gcc (redhat, x86_64, 12.1.1)
 Garbage collector: Serial GC
[2/7] Performing analysis...  [*****]
```

I don't see presence of such lines:
```
 1 user-specific feature(s)
 - org.example.feature.graal.HelloNativeFeature
```

Could you double-check if we're not missing that comparing to my original output?

**Comment by andrejusc** [comment]
Verified this issue now with new available dev build:
https://github.com/graalvm/graalvm-ce-dev-builds/releases/tag/22.3.0-dev-20220901_0249

and using such command to build native image within `nativemain` folder of provided use case after `mvn clean package` command:
```
native-image --verbose --features="org.example.feature.graal.HelloNativeFeature" -o service -p target/service.jar:target/libs -m nativemain/org.example.main.HelloNative 2>&1 | tee ~/hellonative_ni_20220901_1.txt
```
Part of related output is such:
```
========================================================================================================================
GraalVM Native Image: Generating 'service' (executable)...
========================================================================================================================
*** HelloNativeFeature ModuleLayer.boot().modules().size(): 55
...
[1/7] Initializing...                                                                                    (7.0s @ 0.18GB)
 Version info: 'GraalVM 22.3.0-dev Java 17 CE'
 Java version info: '17.0.4+7-jvmci-22.3-b02'
 C compiler: gcc (redhat, x86_64, 8.5.0)
 Garbage collector: Serial GC
 1 user-specific feature(s)
 - org.example.feature.graal.HelloNativeFeature
[2/7] Performing analysis...  [*****]                                                                   (19.6s @ 0.48GB)
```

And when invoking produced executable `service` - it presents now expected output:
```
$ ./service
=== Hello native!
=== ClientConfig: org.glassfish.jersey.client.ClientConfig@a5f7a391
=== JerseyCaller: org.example.feature.JerseyCaller@3dbfd477
```

So with this use case we are good now. But this uncovered some other one, for which I need to write another issue.
